### ISSUE TITLE
port address already in use

### ISSUE THREAD

**Issue by guojunhao118** [cosmtrek/air#216]
![WX20211125-105752](https://user-images.githubusercontent.com/21166467/143372710-d67c1e31-0ce1-448d-b63d-e22e9a826563.png)
I get this problem in version 1.27.5,and i use mac,how can i to solve this?

**Comment by xiantang** [comment]
could u provide the exmaple that i can reproduce

**Comment by xiantang** [comment]
you can try 1.27.6

**Comment by marekmartins** [comment]
I have exactly the same problem using it under WSL2 in Win11. Problem started with version 1.27.5. Also 1.27.6 is affected aswell.

For reproducing dummy HTTP server:
`
package main

import (
	"log"
	"net/http"
)

func main() {
	log.Fatal(http.ListenAndServe(":8080", nil))
}
`
![image](https://user-images.githubusercontent.com/55179666/143444181-d32ab6c8-8a38-4cd4-9af7-0d2e5292fba0.png)

**Comment by xiantang** [comment]
could u use air -d and provide log？

marekmartins ***@***.***>于2021年11月25日 周四20:45写道：

**Comment by marekmartins** [comment]
Sure
```
[crypter] event: "/app/main.go": WRITE
[crypter] event: "/app/main.go": WRITE
[crypter] main.go has changed
[crypter] main.go has changed
[crypter] flushing events
[crypter] main.go has changed
[crypter] building...
[crypter] trying to kill pid 119, cmd [/bin/sh -c /app/tmp/main]
[crypter] running...
[crypter] running process pid 229
[crypter] 2021/11/26 03:35:44 listen tcp :8080: bind: address already in use
```

**Comment by xiantang** [comment]
I have a windows device, i will reproduce this issue, thanks! @marekmartins

**Comment by xiantang** [comment]
which subsystem did u use. I have use ubuntu to try to reproduce this issue but works good. @marekmartins

**Comment by xiantang** [comment]
![1637941783(1)](https://user-images.githubusercontent.com/34479567/143605521-643a3fc8-ee6c-4115-ad0d-755e550b72d2.png)
windows
![image](https://user-images.githubusercontent.com/34479567/143605675-fccd6792-38cd-4986-aa6d-c1a2431ffd3a.png)
ubuntu

**Comment by marekmartins** [comment]
Ubuntu, are You using WSL2? Version 2 is important :)

In powershell:
![image](https://user-images.githubusercontent.com/55179666/143617250-5ac00a1d-2254-4525-9409-20925ea0caa2.png)

**Comment by xiantang** [comment]
https://github.com/cosmtrek/air/pull/218 can try this version, I will release a new tag

**Comment by xiantang** [comment]
https://github.com/cosmtrek/air/releases/tag/v1.27.7

**Comment by marekmartins** [comment]
Slightly output changed, but still port is not released by the moment it tries to run new process.
```
event: "/app/main.go": WRITE
event: "/app/main.go": WRITE
main.go has changed
main.go has changed
flushing events
main.go has changed
building...
trying to kill pid 1269, cmd [/bin/sh -c /app/tmp/main]
killed process pid 1269 successed
cmd killed, pid: 1269
running...
running process pid 1384
2021/11/27 09:32:58 listen tcp :8080: bind: address already in use
```

**Comment by xiantang** [comment]
![image](https://user-images.githubusercontent.com/34479567/143676750-271b25c6-92ea-4365-af25-df3d0a6b950b.png)
thank u , I can reproduce now!!!

**Comment by xiantang** [comment]
![image](https://user-images.githubusercontent.com/34479567/143676768-52fcc8b7-1394-4b97-b6fd-b1c2201de503.png)
i can reproduce now. thanks

**Comment by xiantang** [comment]
@marekmartins  all in all, in macOS use `(p *Process) Kill()` can kill process and it's child process, but in linux( like ubuntu) use this function will just kill target process, it won't kill child process. 
I find a solution https://stackoverflow.com/questions/22470193/why-wont-go-kill-a-child-process-correctly use kill pgid will correct kill process and it's child. 
I have written a PR and work good in my linux env, I will merge this PR soon but will not release a new version.
plz help me check this fix. @marekmartins 

just clone air proj and use `make install`  thanks!!

**Comment by xiantang** [comment]
https://github.com/cosmtrek/air/pull/219/files

**Comment by marekmartins** [comment]
Tested, looks good for me aswell now! Good job!

**Comment by xiantang** [comment]
thanks closing

**Comment by marekmartins** [comment]
there seems to be still one case what is broken and worked with version 1.27.4

if You want to use `dlv` for debugging with `air` like this - `.air.toml`:
```
[build]
cmd = "go build -gcflags='all=-N -l' -o ./tmp/main ."
full_bin = "dlv exec --accept-multiclient --log --headless --continue --listen :2345 --api-version 2 ./tmp/main"
```

Then seems that it perfectly kills `dlv` process, but not anymore `go` process executed by `dlv`

```
main.go has changed
building...
running...
API server listening at: [::]:2345
2021-11-30T07:12:49Z warning layer=rpc Listening for remote connections (connections are not authenticated nor encrypted)
2021-11-30T07:12:49Z info layer=debugger launching process with args: [./tmp/main]
2021-11-30T07:12:49Z debug layer=debugger continuing
2021/11/30 07:12:49 listen tcp :8080: bind: address already in use
2021-11-30T07:12:49Z error layer=rpc writing response:write tcp 127.0.0.1:2345->127.0.0.1:51906: use of closed network connection
```
I can open new issue aswell

**Comment by xiantang** [comment]
just reopen this issue.

**Comment by xiantang** [comment]
@marekmartins  I will try this in local and reproduce it.

**Comment by tahsinature** [comment]
Use these values in `.air.toml`. It worked for me.

```toml
[build]
  delay = 1000
  kill_delay = 10
```

Rest of things will be as it is.

**Comment by marekmartins** [comment]
@tahsinature are You using version `1.27.8`? For me playing with those values does not help, same problem. :(

**Comment by tahsinature** [comment]
> @tahsinature are You using version `1.27.8`? For me playing with those values does not help, same problem. :(

I am using version `1.17.2`
![image](https://user-images.githubusercontent.com/41298152/144228524-d6471529-a4f3-495d-b1e2-f16520404763.png)

**Comment by marekmartins** [comment]
This problem exists only with versions `>=1.27.5`

**Comment by xiantang** [comment]
seems 1.17.2 is golang version.
I will try to fix this issue the weekend. because working day is too busy to do opensouce.

**Comment by marekmartins** [comment]
any news?

**Comment by denialanderror** [comment]
Any progress on this? Air just won't work at all with dlv at the moment, which is pretty much my whole dev environment.

**Comment by stevegore** [comment]
My workaround at the moment is to kill all the stales processes as part of the `full_bin` command.

Assuming your built binary is `/tmp/main`

```bash
full_bin = "pkill -9 'dlv|main'; dlv exec --accept-multiclient --log --headless --continue --listen :2345 --api-version 2 ./tmp/checkbox"
```

**Comment by RakNoel** [comment]
*BUMP*
Issue seems to be with `send_interrupt = true`

Without any modifications to `air.conf`, on some reloads it seems not to kill the outstanding process on the machine at all, before reloading as `ps -ae | grep main.air` shows a long list of duplicated processes.

Changing to `send_interrupt = false` seems to fix the issue for me

```yaml
System:  Arch Linux
Kernel: x86_64 Linux 5.15.13-arch1-1
Framework: Gin
Run: main.go
Inside docker: no
```

**Comment by RakNoel** [comment]
Looking at https://github.com/cosmtrek/air/commit/2cf5ad52c0bc422dff3506f6ce49be5803d31b0c it seems to be getting fixed soon

**Comment by williambao** [comment]
这个问题在mac下一直存在，当更新代码后，经常性的弹 "panic: listen tcp :9000: bind: address already in use"。
有时候连续保存代码10来次后才能正常重新启动

**Comment by xiantang** [comment]
> 这个问题在mac下一直存在，当更新代码后，经常性的弹 "panic: listen tcp :9000: bind: address already in use"。 有时候连续保存代码10来次后才能正常重新启动

Could u provide a example which can stable reproduce?

**Comment by Nekitosss** [comment]
Issue still persists for combination with dlv. 
macOS 12.0.1
Air 1.27.10, built with Go 1.17.6

I am using Air inside docker image. @stevegore's workaround fixes the issue

**Comment by xiantang** [comment]
> Issue still persists for combination with dlv. macOS 12.0.1 Air 1.27.10, built with Go 1.17.6
> 
> I am using Air inside docker image. @stevegore's workaround fixes the issue

could provide more infomation like air config?

**Comment by taman9333** [comment]
Issue still persists on my mac
Air 1.27.10, built with Go 1.17

here is my air config

```
# Config file for [Air](https://github.com/cosmtrek/air) in TOML format

# Working directory
# . or absolute path, please note that the directories following must be under root
root = "."
tmp_dir = "tmp"

[build]
# Just plain old shell command. You could use `make` as well.
cmd = "go build -o /go/bin/project_name"
# Binary file yields from `cmd`.
bin = "/go/bin/project_name"

# Customize binary.
# This is how you start to run your application. Since my application will works like CLI, so to run it, like to make a CLI call.
full_bin = "/go/bin/project_name run"
# This log file places in your tmp_dir.
log = "air_errors.log"
# Watch these filename extensions.
include_ext = ["go", "yaml", "development"]
# Ignore these filename extensions or directories.
exclude_dir = ["tmp"]
# Exclude specific regular expressions.
exclude_regex = ["_test.go"]
# Exclude unchanged files.
exclude_unchanged = true
# It's not necessary to trigger build each time file changes if it's too frequent.
delay = 1000 # ms

[log]
# Show log time
time = true

[misc]
# Delete tmp directory on exit
clean_on_exit = true

```

**Comment by taman9333** [comment]
@xiantang  Issue still persists on my mac
Air 1.27.10, built with Go 1.17

here is my air config

```
# Config file for [Air](https://github.com/cosmtrek/air) in TOML format

# Working directory
# . or absolute path, please note that the directories following must be under root
root = "."
tmp_dir = "tmp"

[build]
# Just plain old shell command. You could use `make` as well.
cmd = "go build -o /go/bin/project_name"
# Binary file yields from `cmd`.
bin = "/go/bin/project_name"

# Customize binary.
# This is how you start to run your application. Since my application will works like CLI, so to run it, like to make a CLI call.
full_bin = "/go/bin/project_name run"
# This log file places in your tmp_dir.
log = "air_errors.log"
# Watch these filename extensions.
include_ext = ["go", "yaml", "development"]
# Ignore these filename extensions or directories.
exclude_dir = ["tmp"]
# Exclude specific regular expressions.
exclude_regex = ["_test.go"]
# Exclude unchanged files.
exclude_unchanged = true
# It's not necessary to trigger build each time file changes if it's too frequent.
delay = 1000 # ms

[log]
# Show log time
time = true

[misc]
# Delete tmp directory on exit
clean_on_exit = true

```

**Comment by nwadams** [comment]
I am able to reproduce this issue consistently when I make many changes close together. What I am seeing in the logs is multiple concurrent builds where the second one ends up with the port in use error. Is there a way to make sure an ongoing build is killed before a new one starts? As well as potentially killing all processes? That, to me would resolve the issue, or at least mitigate it in a way that would allow this issue to be resolved easier.

**Comment by alexksg** [comment]
any solution to this ?

**Comment by marekmartins** [comment]
Yeah problem is still there, seems to be related that with `delve` there are running multiple processes and killing fails.  Latest working version was `v1.27.4`. So there somehow previous processes were killed before newone launched.

**Comment by thanhtungka91** [comment]
i got the same issue, any suggestion?

**Comment by luizgribeiro** [comment]
It looks like it's creating 2 processes when a change occours.
Here's a log with the `-d` flag on version 1.44.0

```2023-07-07T08:48:32.856-0300    INFO   cmd/root.go:96   arquivo de configuração lido    {"config": "local/config
2023-07-07T08:48:32.857-0300    INFO   cmd/start.go:36  inicializando o serviço{"http": "localhost:3028", "grpc"
2023-07-07T08:48:32.858-0300    INFO   server/http.go:41        Servidor configurado com opções TLS     {"cert_f
{"Name":"db/open","SpanContext":{"TraceID":"871dda34d96c315e0d5c19d34cf4bfbc","SpanID":"79f43564127b8821","Trace
{"Name":"db/migrate","SpanContext":{"TraceID":"871dda34d96c315e0d5c19d34cf4bfbc","SpanID":"efc405aa49dd74a1","Tr
{"Name":"bootstrap","SpanContext":{"TraceID":"871dda34d96c315e0d5c19d34cf4bfbc","SpanID":"75642c556de782c4","Tra
[08:48:40] event: CREATE        "/home/luiz/Dev/customers/pkg/server/4913"
[08:48:40] event: CREATE        "/home/luiz/Dev/customers/pkg/server/http.go"
[08:48:40] pkg/server/http.go has changed
[08:48:40] pkg/server/http.go has changed
[08:48:40] event: REMOVE        "/home/luiz/Dev/customers/pkg/server/http.go~"
[08:48:40] trying to kill pid 87597, cmd [/bin/sh -c /home/luiz/Dev/customers/_build/customers start --config lo
[08:48:40] event: CHMOD         "/home/luiz/Dev/customers/pkg/server/http.go"
[08:48:40] event: REMOVE        "/home/luiz/Dev/customers/pkg/server/4913"
[08:48:40] event: CHMOD         "/home/luiz/Dev/customers/pkg/server/4913"
[08:48:40] event: WRITE         "/home/luiz/Dev/customers/pkg/server/http.go"
[08:48:40] pkg/server/http.go has changed
[08:48:40] event: RENAME        "/home/luiz/Dev/customers/pkg/server/http.go"
[08:48:40] event: CREATE        "/home/luiz/Dev/customers/pkg/server/http.go~"
[08:48:40] building...
[08:48:40] pkg/server/http.go has changed
[08:48:40] building...
[08:48:40] cmd killed, pid: 87597
[08:48:43] running...
[08:48:43] running process pid 87820
2023-07-07T08:48:43.943-0300    INFO   cmd/root.go:96   arquivo de configuração lido    {"config": "local/config
2023-07-07T08:48:43.944-0300    INFO   cmd/start.go:36  inicializando o serviço{"http": "localhost:3028", "grpc"
2023-07-07T08:48:43.945-0300    INFO   server/http.go:41        Servidor configurado com opções TLS     {"cert_f
[08:48:43] running...
[08:48:43] running process pid 87828
```

**Comment by ThePrimeagen** [comment]
i also have been able to repro with a simple echo server binding to port 42069 and making any change to code (popos)

**Comment by shanehughes1990** [comment]
@ThePrimeagen I have been having the same issue reloading a gqlgen server fixing it for me seems to be changing build.delay from 0 => 1 in the .air.toml file

I have successfully been able to hot reload every time since.

```
root = "."
testdata_dir = "testdata"
tmp_dir = "tmp"

[build]
  args_bin = []
  bin = "./tmp/main"
  cmd = "go build -o ./tmp/main ."
  delay = 1
  exclude_dir = ["assets", "tmp", "vendor", "testdata"]
  exclude_file = []
  exclude_regex = ["_test.go"]
  exclude_unchanged = false
  follow_symlink = false
  full_bin = ""
  include_dir = []
  include_ext = ["go", "tpl", "tmpl", "html"]
  include_file = []
  kill_delay = "0s"
  log = "build-errors.log"
  poll = false
  poll_interval = 0
  rerun = false
  rerun_delay = 500
  send_interrupt = false
  stop_on_error = false

[color]
  app = ""
  build = "yellow"
  main = "magenta"
  runner = "green"
  watcher = "cyan"

[log]
  main_only = false
  time = false

[misc]
  clean_on_exit = false

[screen]
  clear_on_rebuild = false
  keep_scroll = true

```

**Comment by ramialkaro** [comment]
work around 

```
lsof -i tcp:8080


COMMAND   PID        USER            FD   TYPE     DEVICE    SIZE/OFF   NODE    NAME
main              36407   username     xz    IPv6.     xxxxxxxxx    xxx            TCP       *:8080 (LISTEN)


kill -9 36407
```

**Comment by xiantang** [comment]
https://github.com/cosmtrek/air/commit/bb7fe2887750390eee13daa25348723a5fb20308 I think it's related to this change

**Comment by suntong** [comment]
> could u provide the example that i can reproduce

So far all people complain about the issue but no body was able to produce a minimum example. I'm having exactly the same problem every time I made a change, but found it is hard to produce a minimum example, as I tried with the simplest http server and it works all the time. Anyway,

Here is what I ended up with, an example that is tiny enough to reproduce the problem:

- clone https://github.com/suntong/fiber_demo
- do `go mod tidy`
- Then use Air inside docker image with:

`docker run -it ... -e PORT=8899 -e GOFLAGS="-buildvcs=false" -p 8899:8899 cosmtrek/air`

To trigger the problem:

`echo | tee -a main.go support.go`

and will get:

```
main.go has changed
main.go has changed
building...
building...
running...
running...

2023/10/06 19:48:28 failed to listen: listen tcp4 :8899: listen: address already in use
 âââââââââââââââââââââââââââââââââââââââââââââââââââââ                   Fiber v2.49.2                   â 
 â               http://127.0.0.1:8899               â 
```

@xiantang, I am able to reproduce this issue consistently whenever I make any changes to my files, so hope with this reproducible example, you can fix it soon, as it's been a while since it is reported. 

@luizgribeiro's comment looks to me the closest to the true of the cause of the problem:

> It looks like it's creating 2 processes when a change occours. Here's a log with the `-d` flag on version 1.44.0

**Comment by suntong** [comment]
it is not only creating 2 processes, but can be many:

![image](https://github.com/cosmtrek/air/assets/422244/7e342303-722a-4212-bf9c-4dd743da9207)

**Comment by xiantang** [comment]
thank your minimum example. I have more time now, will look into it

**Comment by xiantang** [comment]
I think this issue should be fixed thanks

**Comment by klausbreyer** [comment]
<img width="1016" alt="Screenshot 2024-02-19 at 13 13 55" src="https://github.com/cosmtrek/air/assets/32771/1fe3a3cd-4044-4344-aa4b-88039f85e64d">

For me the issue was that air was running multiple times in the background. 
It seems it was not wholly killed when I was stopping my make command (that starts air)

Now I try to  use this combination (together with a delay of 1000)

```
kill:
	killall air
	lsof -t -i tcp:8080 | xargs kill -9

start:
	make kill
        air
        .. whatever else
```

**Comment by zeevallin** [comment]
While this is getting fixed, this is my workaround.
I ensure I reap the process with a given listening port before every reload and when you quit `air`.

```toml
post_cmd = [
  "lsof -i :8080 -sTCP:LISTEN | sed 1d | awk '{print $2}' | xargs kill -9"
]
pre_cmd = [
  "lsof -i :8080 -sTCP:LISTEN | sed 1d | awk '{print $2}' | xargs kill -9"
]
```

This part of the `lsof` command is particularly important as it prevents your browser or other tools with open or establishing connections to your server also getting reaped.

```bash
lsof -sTCP:LISTEN
```

Hope it helps and I hope this will be fixed soon so we don't have to use a workaround. ❤️

**Comment by gregawoods** [comment]
I'm experiencing this issue on macOS with a pretty vanilla app using gin. My `.air.toml` file is stock from the`air init` command.

I start seeing the `[ERROR] listen tcp :8080: bind: address already in use` message in my terminal, notice changes are not being picked up, then I have to manually `kill` the process.

**Comment by vladdoroniuk** [comment]
Having the same problem as @gregawoods describes

**Comment by JeetChhatrala** [comment]
Hey @gregawoods & @vladdoroniuk try

```
[misc]
clean_on_exit = true
```

in your `.air.toml` file, before starting with this new config, do manually stop the app's process.

**Comment by gsmatheus** [comment]
I'm having the same issue. Everything was working fine until I restarted my computer. I'm not sure if it's related. I made all the suggested adjustments above but with no success. My version of Air is **v1.52.3**, built with Go **go1.22.5**.

**Comment by alwyn** [comment]
On macos I get this issue when I have my app and proxy ports swapped.  Anyone getting this issue, check if this is not the case for you.

**Comment by rodrigorodriguescosta** [comment]
I still have the issue, I've tried all options above with no luck, my version is ` v1.61.1-5-g7019f7a, built with Go 1.23.1`

is there any workaround? I'm using WSL2 with ubuntu 20.04
